### PR TITLE
Patch freeab

### PR DESCRIPTION
**Pull request by viercc**
There were a few bugs in the implementation of FreeAbelianGroup.

This PR fixes them, and at the same time (sorry!) make FreeAbelianGroup
an abstract type to keep users away from creating invalid values.

### CODE REVIEW

**Review by emilypi — COMMENTED** [review]
Yeah, no, this makes a ton of sense to me, thank you. I've been wondering how best it would be to deal with sparsity and whether we want to represent `mempty`'s (if at all). This solves this really nicely.

File: src/Data/Group/Free.hs [hunk 1]
```diff
@@ -139,18 +147,71 @@ present = flip ($)
...
+    (MkFreeAbelianGroup g) <> (MkFreeAbelianGroup g') =
+      MkFreeAbelianGroup $ mergeG g g'
+      where
+        mergeG = Map.merge
+          Map.preserveMissing
+          Map.preserveMissing
+          (Map.zipWithMaybeMatched $ \_ m n -> nonZero $ m + n)
+        nonZero n = if n == 0 then Nothing else Just n
+    
+    stimes = gtimes
```
Comment (emilypi) [inline src/Data/Group/Free.hs:201]: may as well just use `pow` here and avoid the indirection

File: src/Data/Group/Free/Church.hs [hunk 2]
```diff
@@ -106,16 +107,36 @@ presentFG = flip ($)
...
-  invert (FA g) = FA (invert . g)
+  invert g = pow g (-1 :: Integer)
+
+  {-
+  Note: This implementation "optimizes" from the default implementation of
+  'pow', or more natural
+  
+  > pow (FA g) n = FA $ \k -> gtimes n (g k)
+  
+  by dealying the call of 'gtimes' as late as possible.
```
Comment (emilypi) [inline src/Data/Group/Free/Church.hs:128]:
```suggestion
  by delaying the call of 'gtimes' as late as possible.
```

**Review by emilypi — COMMENTED** [review]
Great! Thank you, this looks like a big step forward for usability. I've made 2 substantive comments, and my only ask is that, since we're hiding `FreeAbelian`'s constructor behind a smart cons, let's expose it in an internal module. Ed's convinced me this is good practice with [this post](https://www.reddit.com/r/haskell/comments/hwge1o/haskeller_competency_matrix/fz2zvyt/).

File: src/Data/Group/Free.hs (hunk 1 above)
Comment (viercc) [reply src/Data/Group/Free.hs:201]:
```suggestion
    stimes = flip pow
```

**Review by emilypi — APPROVED** [review]

**Comment by emilypi** [comment]
Brilliant! thanks @viercc

**Comment by emilypi** [comment]
Actually, now that i see it, the changes won't compile because you need `a` in principle, so I'm fine with keeping the typeclasses.

**Pull request merged**